File: algorithms/graphs/src/bellman_ford.cpp
#include "bellman_ford.hpp"

const int INF = 1000000000;

std::vector<int> bellmanFord(
    const std::vector<std::vector<std::pair<int, int>>> &graph,
    int source
) {
    std::vector<int> distances(graph.size(), INF);
    distances[source] = 0;

    for (int i = 0; i < graph.size() - 1; i++) {
        for (int vtx = 0; vtx < graph.size(); vtx++) {
            if (distances[vtx] == INF) {
                continue;
            }

            for (const auto& edges : graph[vtx]) {
                int neighbourVtx = edges.first;
                int vtxToNeighbourDist = edges.second;

                if (distances[vtx] + vtxToNeighbourDist < distances[neighbourVtx]) {
                    distances[neighbourVtx] = distances[vtx] + vtxToNeighbourDist;
                }
            }
        }
    }

    return distances;
}

File: algorithms/graphs/python/bellman_ford.py


def main():
    n, m = map(int, input().split())
    graph = [[] for _ in range(n)]

    for _ in range(m):
        v_from, v_to, weight = map(int, input().split())
        graph[v_from].append([v_to, weight])
    
    source, target = map(int, input().split())

    distances = [float('inf') for _ in range(n)]
    distances[source] = 0

    for _ in range(n - 1):
        for vtx in range(n):
            if distances[vtx] == float("inf"):
                continue

            for vtx_to, weight in graph[vtx]:
                updated_dist = distances[vtx] + weight
                if updated_dist < distances[vtx_to]:
                    distances[vtx_to] = updated_dist
    
    print(distances)


if __name__ == "__main__":
    main()
